README.md
[![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](LICENSE.txt)
[![Webex Events](https://github.com/SocioEvents/webex-events-go-sdk/actions/workflows/go-test.yml/badge.svg)](https://github.com/SocioEvents/webex-events-go-sdk/actions)

⚠️ This library has not been released yet.
# Webex Events Api Go SDK

Webex Events provides a range of additional SDKs to accelerate your development process.
They allow a standardized way for developers to interact with and leverage the features and functionalities.
Pre-built code modules will help access the APIs with your private keys, simplifying data gathering and update flows.

Requirements
-----------------

Go 1.21+

Installation
-----------------

`go get github.com/SocioEvents/webex-events-go-sdk`

Configuration
-----------------

```go
	var ctx = context.Background()
	config := NewConfig()
	config.SetAccessToken("sk_live_your_access_token")
	config.SetMaxRetries(5) // Default is 5
	config.SetTimeout(time.Duration(30) * time.Second) // default is 30 seconds
	var loggerHandler = slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelInfo})
	config.SetLoggerHandler(loggerHandler) // default is Error loglevel to stdout
```

Usage
-----------------
```go
	var headers = http.Header{}

	var variables = make(map[string]any)
	var query = "query CurrenciesList{ currenciesList{ isoCode }}"
	var operationName = "CurrenciesList"

	queryRequest := QueryRequest{
		Query:         query,
		OperationName: operationName,
		Variables:     variables,
		Headers:       headers,
	}
	var response, err = client.Query(ctx, &queryRequest)
	fmt.Println(response, err)
```

For Introspection
-----------------
```go
	var response, err = client.DoIntrospectionQuery(ctx)
	fmt.Println(response, err)
```

Idempotency
-----------------
The API supports idempotency for safely retrying requests without accidentally performing the same operation twice.
When doing a mutation request, use an idempotency key. If a connection error occurs, you can repeat
the request without risk of creating a second object or performing the update twice.

To perform mutation request, you must add a header which contains the idempotency key such as
`Idempotency-Key: <your key>`. The SDK does not produce an Idempotency Key on behalf of you if it is missed.
The SDK also validates the key on runtime, if it is not valid UUID token it will raise an exception. Here is an example
like the following:

```go
	var headers = http.Header{}
	headers.Set("Idempotency-Key", "61672155-56d3-4375-a864-52e7bba4f445") // This is only for mutations.

	var variables = map[string]any{
		"input": map[string]any{
			"ids":     []int{1, 2, 3},
			"eventId": 1,
		},
	}
	var query = `
          mutation TrackDelete($input: TrackDeleteInput!) {
            trackDelete(input: $input) {
              success
            }
          }
`
	var operationName = "TrackDelete"

	queryRequest := QueryRequest{
		Query:         query,
		OperationName: operationName,
		Variables:     variables,
		Headers:       headers,
	}
	var response, err = client.Query(ctx, &queryRequest)
	fmt.Println(response, err)
```

Telemetry Data Collection
-----------------
Webex Events collects telemetry data, including hostname, operating system, language and SDK version, via API requests.
This information allows us to improve our services and track any usage-related faults/issues. We handle all data with
the utmost respect for your privacy. For more details, please refer to the Privacy Policy at https://www.cisco.com/c/en/us/about/legal/privacy-full.html

Development
-----------------

After checking out the repo, install dependencies. Then, run `go test` to run the tests.

Contributing
-----------------
Please see the [contributing guidelines](CONTRIBUTING.md).

License
-----------------

The library is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

Code of Conduct
-----------------

Everyone interacting in the Webex Events API project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/SocioEvents/webex-events-go-sdk/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [Unreleased]

## [0.1.0] - 2023-11-25

- Initial release

client_test.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"os"
	"runtime"
	"strings"
	"testing"
)

type TestHttpClient struct {
	mock.Mock
}

func (c *TestHttpClient) Do(req *http.Request) (*http.Response, error) {
	args := c.Called(req)
	return args.Get(0).(*http.Response), args.Error(1)
}

func TestWhenAccessTokenIsBlank(t *testing.T) {
	ctx := context.Background()
	config := NewConfig()
	client := NewClient(config)

	var variables = make(map[string]any)
	headers := http.Header{}
	var query = "{ currenciesList{ isoCode }}"
	var operationName = "CurrenciesList"

	response, err := client.Query(ctx, &QueryRequest{
		Query:         query,
		OperationName: operationName,
		Variables:     variables,
		Headers:       headers,
	})

	assert.Nil(t, response)
	assert.ErrorIs(t, err, AccessTokenRequiredError)
}

func TestQueryWith409StatusCode(t *testing.T) {
	ctx := context.Background()
	config := NewConfig()
	config.SetAccessToken("sk_live_some_access_token")
	config.SetMaxRetries(3)
	config.SetLoggerHandler(slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelInfo}))
	client := NewClient(config)

	var headers = http.Header{}
	headers.Set("Idempotency-Key", "61672155-56d3-4375-a864-52e7bba4f445")
	var variables = make(map[string]any)

	var query = "{ currenciesList{ isoCode }}"
	var operationName = "CurrenciesList"

	var responseBody = `{ "message": "Conflict", "extensions": { code: "CONFLICT" }}`

	httpClient := new(TestHttpClient)
	stringReader := strings.NewReader(responseBody)
	body := io.NopCloser(stringReader)

	reqUrl, err := url.Parse("http://localhost")
	assert.NoError(t, err)

	header := http.Header{
		"Authorization":      []string{fmt.Sprintf("Bearer %s", config.GetAccessToken())},
		"User-Agent":         []string{"Webex Go SDK"},
		"Content-Type":       []string{"application/json"},
		"X-Sdk-Name":         []string{"Go SDK"},
		"X-Sdk-Version":      []string{VERSION},
		"X-Sdk-Lang-Version": []string{runtime.Version()},
		"Idempotency-Key":    []string{"61672155-56d3-4375-a864-52e7bba4f445"},
	}
	var mockResponse = &http.Response{
		Request: &http.Request{
			Header: header,
			URL:    reqUrl,
		},
		Status:     "409",
		StatusCode: 409,
		Body:       body,
	}

	httpClient.On("Do", mock.Anything).Return(mockResponse, nil)
	client.SetHttpClient(httpClient)

	response, err := client.Query(ctx, &QueryRequest{
		Query:         query,
		OperationName: operationName,
		Variables:     variables,
		Headers:       headers,
	})

	assert.Equal(t, 2, response.RetryCount)
}

func TestQueryWith200StatusCode(t *testing.T) {
	ctx := context.Background()
	config := NewConfig()
	config.SetAccessToken("sk_live_some_access_token")
	client := NewClient(config)

	var headers = http.Header{}
	headers.Set("Idempotency-Key", "61672155-56d3-4375-a864-52e7bba4f445")
	var variables = make(map[string]any)

	var query = "{ currenciesList{ isoCode }}"
	var operationName = "CurrenciesList"

	var responseBody = `{"data": { "currenciesList": { "isoCode": "USD" }}}`

	httpClient := new(TestHttpClient)
	stringReader := strings.NewReader(responseBody)
	body := io.NopCloser(stringReader)

	reqUrl, err := url.Parse("http://localhost")
	assert.NoError(t, err)

	header := http.Header{
		"Authorization":      []string{fmt.Sprintf("Bearer %s", config.GetAccessToken())},
		"User-Agent":         []string{"Webex Go SDK"},
		"Content-Type":       []string{"application/json"},
		"X-Sdk-Name":         []string{"Go SDK"},
		"X-Sdk-Version":      []string{VERSION},
		"X-Sdk-Lang-Version": []string{runtime.Version()},
		"Idempotency-Key":    []string{"61672155-56d3-4375-a864-52e7bba4f445"},
	}
	var mockResponse = &http.Response{
		Request: &http.Request{
			Header: header,
			URL:    reqUrl,
		},
		Status:     "200",
		StatusCode: 200,
		Body:       body,
		Header: http.Header{
			"Content-Type":           []string{"application/json"},
			"X-Daily-Call-Limit":     []string{"10/200"},
			"X-Secondly-Call-Limit":  []string{"5/50"},
			"X-Daily-Retry-After":    []string{"3"},
			"X-Secondly-Retry-After": []string{"5"},
		},
	}

	httpClient.On("Do", mock.Anything).Return(mockResponse, nil)
	client.SetHttpClient(httpClient)

	response, err := client.Query(ctx, &QueryRequest{
		Query:         query,
		OperationName: operationName,
		Variables:     variables,
		Headers:       headers,
	})
	assert.NoError(t, err)
	assert.NotNil(t, response)
	result := make(map[string]any)
	err = json.Unmarshal([]byte(response.RequestBody), &result)
	assert.NoError(t, err)
	assert.EqualValues(t, result["operationName"], operationName)
	assert.EqualValues(t, result["query"], query)
	assert.EqualValues(t, result["variables"], variables)
	assert.NoError(t, err)
	assert.EqualValues(t, response.Status, http.StatusOK)
	assert.EqualValues(t, response.Body, responseBody)
	assert.EqualValues(t, response.RequestHeaders["Authorization"][0], "Bearer "+config.GetAccessToken())
	assert.Contains(t, response.RequestHeaders["User-Agent"][0], "Webex Go SDK")
	assert.EqualValues(t, response.RequestHeaders["Content-Type"][0], "application/json")
	assert.EqualValues(t, response.RequestHeaders["X-Sdk-Name"][0], "Go SDK")
	assert.EqualValues(t, response.RequestHeaders["X-Sdk-Version"][0], VERSION)
	assert.EqualValues(t, response.RequestHeaders["X-Sdk-Lang-Version"][0], runtime.Version())
	assert.EqualValues(t, response.RequestHeaders["Idempotency-Key"][0], "61672155-56d3-4375-a864-52e7bba4f445")
}


helpers_test.go
package main

import (
	"github.com/stretchr/testify/assert"
	_ "github.com/stretchr/testify/assert"
	"strings"
	"testing"
)

func Test_getUserAgent(t *testing.T) {
	var got = getUserAgent()
	var want = strings.HasPrefix(got, "Webex Go SDK")

	assert.EqualValues(t, want, true)

	// From cache

	got = getUserAgent()
	want = strings.HasPrefix(got, "Webex Go SDK")
	assert.EqualValues(t, want, true)
}

func TestGetRequestUrl(t *testing.T) {
	config := NewConfig()
	config.SetAccessToken("sk_test_some_access_token1")
	client := NewClient(config)
	var got = client.getRequestUrl()
	var want = "https://public.sandbox-api.socio.events/graphql"

	assert.EqualValues(t, got, want)

	// Retry again. The second call will retrieve from cache.
	got = client.getRequestUrl()
	want = "https://public.sandbox-api.socio.events/graphql"

	assert.EqualValues(t, got, want)

	config = NewConfig()
	config.SetAccessToken("sk_live_some_access_token1")
	client = NewClient(config)

	got = client.getRequestUrl()
	want = "https://public.api.socio.events/graphql"

	assert.EqualValues(t, got, want)
}

func Test_fillErrorResponseIfStatusCodeIsInRange(t *testing.T) {
	var json = `
	{
		"message": "Something went wrong",
		"extensions": {
			"code": "TOKEN_IS_EXPIRED"
		}
	}
`
	config := NewConfig()
	config.SetAccessToken("sk_live_some_access_token")
	client := NewClient(config)

	errorResponse, err := client.errorResponse([]byte(json))
	assert.NoError(t, err)
	assert.EqualValues(t, errorResponse.Message, "Something went wrong")

}

func Test_fillErrorResponseIfJSONisInvalid(t *testing.T) {
	var json = "malformed json"
	config := NewConfig()
	config.SetAccessToken("sk_live_some_access_token")
	client := NewClient(config)

	_, err := client.errorResponse([]byte(json))
	assert.Error(t, err, BadResponseError{
		response: []byte("malformed json"),
	})
	//assert.Errorf(t, err, "The provided JSON is not valid. Provided body is: malformed json")
}

func Test_fillErrorResponseWithAllValues(t *testing.T) {
	var json = `
		{
			"message": "Something went wrong",
			"extensions": {
				"code": "TOKEN_IS_EXPIRED",
				"referenceId": "reference id",
				"errors": {
					"first_name": [
						"invalid"
					]
				},
				"cost": 51,
				"availableCost": 50,
				"threshold": 50,
				"dailyThreshold": 200,
				"dailyAvailableCost": 190
			},
			"errors": [
				{
					"message": "graphql error"
				}
			]
		}
`
	config := NewConfig()
	config.SetAccessToken("sk_live_some_access_token")
	client := NewClient(config)

	errorResponse, err := client.errorResponse([]byte(json))
	assert.NoError(t, err)
	assert.EqualValues(t, errorResponse.Message, "Something went wrong")
	assert.EqualValues(t, errorResponse.Extensions.Code, "TOKEN_IS_EXPIRED")
	assert.EqualValues(t, errorResponse.Extensions.ReferenceId, "reference id")
	assert.EqualValues(t, errorResponse.Extensions.Cost, 51)
	assert.EqualValues(t, errorResponse.Extensions.AvailableCost, 50)
	assert.EqualValues(t, errorResponse.Extensions.Threshold, 50)
	assert.EqualValues(t, errorResponse.Extensions.DailyThreshold, 200)
	assert.EqualValues(t, errorResponse.Extensions.DailyAvailableCost, 190)

	_, ok := errorResponse.Extensions.Errors["first_name"]
	assert.EqualValues(t, ok, true)

	msg, ok := errorResponse.GraphqlErrors[0]["message"]
	assert.EqualValues(t, ok, true)
	assert.EqualValues(t, msg, "graphql error")
}


rate_limiter_test.go
package main

import (
	"github.com/stretchr/testify/assert"
	_ "github.com/stretchr/testify/assert"
	"net/http"
	"testing"
)

func Test_fillRateLimiter(t *testing.T) {
	var mockResponse = http.Response{
		Header: http.Header{
			"Content-Type":           []string{"application/json"},
			"X-Daily-Call-Limit":     []string{"10/200"},
			"X-Secondly-Call-Limit":  []string{"5/50"},
			"X-Daily-Retry-After":    []string{"3"},
			"X-Secondly-Retry-After": []string{"5"},
		},
	}

	var rateLimiter = RateLimiter{}
	rateLimiter.fill(&mockResponse)
	assert.Equal(t, rateLimiter.DailyBasedCostThreshold, 200)
	assert.Equal(t, rateLimiter.UsedDailyBasedCost, 10)
	assert.Equal(t, rateLimiter.SecondBasedCostThreshold, 50)
	assert.Equal(t, rateLimiter.UsedSecondBasedCost, 5)
	assert.Equal(t, rateLimiter.DailyRetryAfterInSecond, 3)
	assert.Equal(t, rateLimiter.SecondlyRetryAfterInMs, 5)
}


helpers.go
package main

import (
	"fmt"
	"os"
	"runtime"
)

var (
	userAgent          string
	introspectionQuery string
)

func getIntrospectionQuery() string {
	if len(introspectionQuery) > 0 {
		return introspectionQuery
	} else {
		return `
                query IntrospectionQuery {
                  __schema {
                   s
                    queryType { name }
                    mutationType { name }
                    subscriptionType { name }
                    types {
                      ...FullType
                    }
                    directives {
                      name
                      description
                      locations
                     s
                      args {
                        ...InputValue
                      }
                    }
                  }
                }
                fragment FullType on __Type {
                  kind
                  name
                  description
                 s
                 s
                  fields(includeDeprecated: true) {
                    name
                    description
                    args {
                      ...InputValue
                    }
                    type {
                      ...TypeRef
                    }
                    isDeprecated
                    deprecationReason
                  }
                  inputFields {
                    ...InputValue
                  }
                  interfaces {
                    ...TypeRef
                  }
                  enumValues(includeDeprecated: true) {
                    name
                    description
                    isDeprecated
                    deprecationReason
                  }
                  possibleTypes {
                    ...TypeRef
                  }
                }
                fragment InputValue on __InputValue {
                  name
                  description
                  type { ...TypeRef }
                  defaultValue
                 s
                 s
                }
                fragment TypeRef on __Type {
                  kind
                  name
                  ofType {
                    kind
                    name
                    ofType {
                      kind
                      name
                      ofType {
                        kind
                        name
                        ofType {
                          kind
                          name
                          ofType {
                            kind
                            name
                            ofType {
                              kind
                              name
                              ofType {
                                kind
                                name
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
`
	}
}

func getUserAgent() string {
	if len(userAgent) > 0 {
		return userAgent
	}

	hostname, _ := os.Hostname()
	userAgent = fmt.Sprintf(
		"Webex Go SDK(v%s) - OS(%s) - hostname(%s) - Go Version(%s)",
		VERSION,
		runtime.GOOS,
		hostname,
		runtime.Version())

	return userAgent
}


rate_limiter.go
package main

import (
	"net/http"
	"strconv"
	"strings"
)

const DailyCallLimit = "X-Daily-Call-Limit"
const SecondlyCallLimit = "X-Secondly-Call-Limit"
const DailyRetryAfter = "X-Daily-Retry-After"
const SecondlyRetryAfter = "X-Secondly-Retry-After"

type RateLimiter struct {
	UsedSecondBasedCost      int
	SecondBasedCostThreshold int
	UsedDailyBasedCost       int
	DailyBasedCostThreshold  int
	DailyRetryAfterInSecond  int
	SecondlyRetryAfterInMs   int
}

func NewRateLimiter() *RateLimiter {
	return new(RateLimiter)
}

func (rateLimiter *RateLimiter) fill(resp *http.Response) {
	var dailyCallLimit = resp.Header.Get(DailyCallLimit)
	if len(dailyCallLimit) > 0 {
		var limit = strings.Split(dailyCallLimit, "/")
		rateLimiter.UsedDailyBasedCost, _ = strconv.Atoi(limit[0])
		rateLimiter.DailyBasedCostThreshold, _ = strconv.Atoi(limit[1])
	}

	var secondlyCallLimit = resp.Header.Get(SecondlyCallLimit)
	if len(secondlyCallLimit) > 0 {
		var limit = strings.Split(secondlyCallLimit, "/")
		rateLimiter.UsedSecondBasedCost, _ = strconv.Atoi(limit[0])
		rateLimiter.SecondBasedCostThreshold, _ = strconv.Atoi(limit[1])
	}

	var dailyRetryAfter = resp.Header.Get(DailyRetryAfter)
	if len(dailyRetryAfter) > 0 {
		rateLimiter.DailyRetryAfterInSecond, _ = strconv.Atoi(dailyRetryAfter)
	}

	var secondlyRetryAfter = resp.Header.Get(SecondlyRetryAfter)
	if len(secondlyRetryAfter) > 0 {
		rateLimiter.SecondlyRetryAfterInMs, _ = strconv.Atoi(secondlyRetryAfter)
	}
}


configuration.go
package main

import (
	"log/slog"
	"os"
	"time"
)

type Config struct {
	accessToken string
	timeout     time.Duration
	maxRetries  uint
	logger      *slog.Logger
}

func NewConfig() *Config {
	var config = &Config{
		timeout: time.Duration(30),
	}

	var handlerOptions = &slog.HandlerOptions{Level: slog.LevelWarn}
	config.logger = slog.New(slog.NewTextHandler(os.Stdout, handlerOptions))
	return config
}

func (c *Config) SetAccessToken(token string) {
	c.accessToken = token
}

func (c *Config) SetTimeout(timeout time.Duration) {
	if timeout < time.Duration(1)*time.Second {
		panic("timeout must be greater than or equal to 1 second")
	}
	c.timeout = timeout
}

func (c *Config) SetMaxRetries(maxRetries uint) {
	c.maxRetries = maxRetries
}

func (c *Config) SetLoggerHandler(handler slog.Handler) {
	if handler == nil {
		var handlerOptions = &slog.HandlerOptions{Level: slog.LevelWarn}
		c.logger = slog.New(slog.NewTextHandler(os.Stdout, handlerOptions))
	} else {
		c.logger = slog.New(handler)
	}
}

func (c *Config) GetAccessToken() string {
	return c.accessToken
}

func (c *Config) GetTimeout() time.Duration {
	return c.timeout
}

func (c *Config) GetMaxRetries() uint {
	return c.maxRetries
}


client.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"runtime"
	"slices"
	"strings"
	"time"
)

const VERSION = "0.1.0"

var AccessTokenRequiredError = errors.New("access token is required")
var RetriableHttpStatuses = []int{408, 409, 429, 502, 503, 504}

type BadResponseError struct {
	response []byte
}

func (err BadResponseError) Error() string {
	return fmt.Sprintf("the provided JSON is not valid. provided body is:%s", string(err.response))
}

type HTTPClient interface {
	Do(req *http.Request) (*http.Response, error)
}

type Client struct {
	config *Config
	client HTTPClient
	url    *string
}

type QueryRequest struct {
	Query         string
	OperationName string
	Variables     map[string]any
	Headers       http.Header
}

func NewClient(config *Config) *Client {
	c := &Client{
		config: config,
	}

	if c.config.GetTimeout() < (time.Duration(1) * time.Second) {
		c.config.SetTimeout(time.Duration(30) * time.Second)
	}
	c.SetHttpClient(&http.Client{Timeout: config.GetTimeout() * time.Second})
	return c
}

func (c *Client) SetHttpClient(httpClient HTTPClient) {
	c.client = httpClient
}

func (c *Client) DoIntrospectionQuery(ctx context.Context) (*Response, error) {
	client := NewClient(c.config)
	var headers = http.Header{}
	var variables = make(map[string]any)
	var r = QueryRequest{
		OperationName: "IntrospectionQuery",
		Headers:       headers,
		Variables:     variables,
		Query:         getIntrospectionQuery(),
	}

	var response, err = client.Query(ctx, &r)
	if err != nil {
		return nil, err
	}

	return response, nil
}

func (c *Client) Query(ctx context.Context, queryRequest *QueryRequest) (*Response, error) {
	if c.config.GetAccessToken() == "" {
		return nil, AccessTokenRequiredError
	}

	var requestBody = make(map[string]any)
	requestBody["operationName"] = queryRequest.OperationName
	requestBody["variables"] = queryRequest.Variables
	requestBody["query"] = queryRequest.Query
	jsonBody, err := json.Marshal(requestBody)
	if err != nil {
		c.config.logger.Error(err.Error())
		return nil, err
	}

	req, err := http.NewRequestWithContext(ctx, "POST", c.getRequestUrl(), bytes.NewBuffer(jsonBody))
	if err != nil {
		c.config.logger.Error(err.Error())
		return nil, err
	}

	req.Header = queryRequest.Headers
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+c.config.GetAccessToken())
	req.Header.Set("X-Sdk-Name", "Go SDK")
	req.Header.Set("X-Sdk-Version", VERSION)
	req.Header.Set("X-Sdk-Lang-Version", runtime.Version())
	req.Header.Set("User-Agent", getUserAgent())

	var (
		start    = time.Now()
		wait     = 250.0
		waitRate = 1.4
		resp     *http.Response
		retries  = uint(0)
	)
	if c.config.maxRetries < 1 {
		c.config.SetMaxRetries(5)
	}

	// Retry loop
	c.config.logger.Info("The request to" + c.getRequestUrl() + " endpoint has been started.")
	for ; retries < c.config.maxRetries; retries++ {
		resp, err = c.client.Do(req)
		if err != nil || slices.Contains(RetriableHttpStatuses, resp.StatusCode) {
			if err != nil {
				c.config.logger.Error("The request is going to be retried. The error message is: " + err.Error())
			} else {
				c.config.logger.Error("The request is going to be retried due to the fact that the server returned " + resp.Status + " status code.")
			}
			wait *= waitRate
			c.config.logger.Info(fmt.Sprintf("Sleeping for %d ms...", int(wait)))
			time.Sleep(time.Duration(wait) * time.Millisecond)
			continue
		}
		break
	}

	if resp != nil && resp.StatusCode > 299 {
		c.config.logger.Error(fmt.Sprintf("The request is failed. The request retried %d times but server returned %d status code", retries, resp.StatusCode))
	}

	if err != nil {
		c.config.logger.Error(err.Error())
		return nil, err
	}
	defer resp.Body.Close()

	var elapsed = time.Since(start)

	if err != nil {
		c.config.logger.Error(err.Error())
		return nil, err
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		c.config.logger.Error(err.Error())
		return nil, err
	}

	var rateLimiter = NewRateLimiter()
	rateLimiter.fill(resp)

	response := &Response{
		Status:         resp.StatusCode,
		Headers:        resp.Header,
		Body:           string(body),
		RequestHeaders: resp.Request.Header,
		RequestBody:    string(jsonBody),
		Url:            resp.Request.URL.String(),
		RetryCount:     int(retries) - 1,
		TimeSpentInMs:  elapsed,
		RateLimiter:    rateLimiter,
	}
	if response.Status >= http.StatusBadRequest && response.Status <= http.StatusInternalServerError {
		errorResponse, err := c.errorResponse(body)
		if err != nil {
			c.config.logger.Error(err.Error())
			return response, err
		}
		response.ErrorResponse = errorResponse
	}

	return response, nil
}

func (c *Client) getRequestUrl() string {
	if c.url != nil {
		return *c.url
	}
	var path = "/graphql"
	var url string
	if strings.HasPrefix(c.config.GetAccessToken(), "sk_live") {
		url = "https://public.api.socio.events"
	} else {
		url = "https://public.sandbox-api.socio.events"
	}
	uri := url + path
	c.url = &uri
	return uri
}

func (c *Client) errorResponse(body []byte) (*ErrorResponse, error) {
	errorResponse := NewErrorResponse()
	if json.Valid(body) {
		err := json.Unmarshal(body, &errorResponse)
		if err != nil {
			return nil, err
		}
		return errorResponse, nil
	} else {
		return nil, BadResponseError{response: body}
	}
}
